src/services/schedule.ts

```diff
@@ -104,41 +84,48 @@ export class ScheduleService implements IScheduleService {
     const isDevTx = this.isDevTx(request.paymentEmail);
     const isValidCouponCode = this.isValidCouponCode(request.paymentRefundAddress);
 
-    const freeTx = isDevTx || isValidCouponCode || !PAYMENTS_ENABLED;
+    const isFreeTx = isDevTx || isValidCouponCode || !PAYMENTS_ENABLED;
 
     const prevStatus = transaction.status;
 
-    if (params?.apiKey) {
-      if (params?.draft === true || (params?.draft as any) === 'true') {
-        transaction.status = transaction.status || Status.Draft;
-      } else {
-        transaction.status = transaction.status === Status.Draft ? Status.Pending : transaction.status;
-      }
-      transaction.status = transaction.status || Status.Pending;
-    } else {
-      transaction.status = freeTx ? Status.Pending : Status.PendingPayment;
-    }
-    transaction.paymentAddress = freeTx ? '' : PaymentService.getNextPaymentAddress();
+    transaction.status = calculateNewStatus({
+      currentStatus: transaction.status,
+      isFreeTx,
+      isDraft: params.draft,
+      isProxyRequest,
+      isStrategyTx,
+    });
+
+    transaction.paymentAddress = isFreeTx ? '' : PaymentService.getNextPaymentAddress();
 
     // extra check for duplicate in db
     // if same tx didn't exist when the process started
     // this should still be the case
     // otherwise we have a problem
     if (!transactionExists) {
-      let duplicate = await this.findBySignedTransaction(request.signedTransaction);
+      const duplicate = await findBySignedTransaction(request.signedTransaction);
```

Comment:
Another nested if. But this time there's a varaible in between making it difficult to simplify. My recommendation is, putting the variable outside might cause lower performance and time consumption tho,

```
const duplicate = await findBySignedTransaction(request.signedTransaction);
if (!transactionExists && duplicate) {
        throw new Error(`Duplicate transaction ${request.signedTransaction}`);
    }
```

Reply:
Correct, this is ugly. The better solution is to extract the inner logic into a separate function, which I did